Introducing Schema Crawler

albertodelazzari opened the pull request:
I refactored the code in order to use Schema Crawler. 
These are the key points:

- common interfaces to database objects (no more QueryResults for DatabaseClient)
- easy filtering for table exclusions or other database objects
- easy support for other databases (I'm working on Microsoft SQL server support)

I also added this changes:

- SqlQuote are now no more hard-coded but automatically loaded according to MetaData information (table/column quote char, constant char and catalog separator)
- I introduced DatabaseCatalog an abstraction layer over Schema Crawler that is used by TableInfoAssembler
- TableInfoAssembler has no references to raw sql row results anymore
- All the methods I was able to delete (because they became useless) were removed the other ones were deprecated
- Schema name for Oracle integration tests are upper case (because it's the default behaviour)
- I changed all the GRANT(s) for MySql scripts from "localhost" to "%" in order to avoid problems with docker instances

 Things to do:

- remove deprecated methods
- review inverted exclusion of tables in order to use the current implementation through SchemaCrawler inclusion rules
- Column exclusion according to its type (blob, binary, etc) should be reviewed according to SchemaCrawler inclusion rules
- DatabaseType probably could be removed
- SqlDatatype could be removed as well (with SchemaCrawler we can have the corresponding Java type for each column)
- According to different case support for each database (case sensitive or case insensitive) when the specified schema can't be found by SchemaCrawler we need to notify the user that should pay attention to this

@jexp take a look at these changes and give me feedbacks, thanks.

jexp commented:
@albertodelazzari looks good to me, please go ahead with your remaining suggestions.

jexp commented:
@albertodelazzari let me know when this is ready to be merged.

albertodelazzari commented:
I added the following changes:

- removed SQL Quotes where they are  not useful or hard coded. There is no need to have them because they are all dynamically built from Metadata information

- No more DB specific data types. These are now abstracted by JavaSqlType from SchemaCrawler. SqlDataType is not more an Enum but a simple plain java class. For now there are only 3 default types (INT, TEXT and TINYINT) just to simplify tests

- Neo4j data types has been reviewed in order to be synced with Neo4j Java Driver types

- No more class name for the JDBC drivers. Class.forName is useless since JDBC 4 version so we don't need to specify this string in DatabaseClient and don't need to call forName method.

- Deprecated methods removed

I'm working on:

- Column exclusion according to its type (blob, binary, etc) should be reviewed according to SchemaCrawler inclusion rules

- review inverted exclusion of tables in order to use the current implementation through SchemaCrawler inclusion rules

After that the PR could be considered as complete

jexp commented:
Sounds good to me. 

Some older jdbc drivers might still be on jdbc3. So I wonder if we should leave the driver class name as emergency fallback 

Von meinem iPhone gesendet

> Am 20.06.2017 um 11:25 schrieb Alberto De Lazzari <notifications@github.com>:
> 
> I added the following changes:
> 
> removed SQL Quotes where they are not useful or hard coded. There is no need to have them because they are all dynamically built from Metadata information
> 
> No more DB specific data types. These are now abstracted by JavaSqlType from SchemaCrawler. SqlDataType is not more an Enum but a simple plain java class. For now there are only 3 default types (INT, TEXT and TINYINT) just to simplify tests
> 
> Neo4j data types has been reviewed in order to be synced with Neo4j Java Driver types
> 
> No more class name for the JDBC drivers. Class.forName is useless since JDBC 4 version so we don't need to specify this string in DatabaseClient and don't need to call forName method.
> 
> Deprecated methods removed
> 
> I'm working on:
> 
> Column exclusion according to its type (blob, binary, etc) should be reviewed according to SchemaCrawler inclusion rules
> 
> review inverted exclusion of tables in order to use the current implementation through SchemaCrawler inclusion rules
> 
> After that the PR could be considered as complete
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jexp commented:
Looked at the code, looks good to me. 
@albertodelazzari what do you think about my comment?

albertodelazzari commented:
@jexp ok for the emergency fallback I just only want to do a fast check if we can avoid to define the fully qualified class name (for each driver) in the DatabaseType class.

It will be nice to have the db type parameter removed from the CLI

jexp commented:
Yes no need for the type class

Von meinem iPhone gesendet

> Am 21.06.2017 um 16:11 schrieb Alberto De Lazzari <notifications@github.com>:
> 
> @jexp ok for the emergency fallback I just only want to do a fast check if we can avoid to define the fully qualified class name (for each driver) in the DatabaseType class.
> 
> It will be nice to have the db type parameter removed from the CLI
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

albertodelazzari commented:
@jexp the last two open points were closed. I also modified the DatabaseInspector in order to accept FilterOptions as a parameter in order to have more flexibility in filtering (for example, not only filtering by column type but also by column name).
